### PR TITLE
Remote Signer Server Full Implementation

### PR DESCRIPTION
**Pull request by rauljordan**
This PR is part of #1, it implements most remote-signer functionality with tests and godocs. The aim of this project is to serve as a solid reference for how to build a remote keysigner for Prysm, and we aim to keep it to a high standard for users to look at.

### CODE REVIEW

File: keyvault/deterministic/store.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
+/*
+Package deterministic defines an implementation of an insecure,
+purely deterministic keyvault used by this remote signer server
+for demonstration purposes.
+
+WARN: This is a toy implementation that returns deterministic BLS12-381
+secret+public key pairs, NOT MEAN FOR PRODUCTION.
```
Comment (farazdagi) [inline keyvault/deterministic/store.go]:
```suggestion
secret+public key pairs, NOT MEANT FOR PRODUCTION.
```

File: rpc/remote_signer_test.go [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+	"bytes"
+	"context"
+	"errors"
+	"strings"
+	"testing"
+
+	ptypes "github.com/gogo/protobuf/types"
+	validatorpb "github.com/prysmaticlabs/prysm/proto/validator/accounts/v2"
+	"github.com/prysmaticlabs/prysm/shared/bls"
+
```
Comment (farazdagi) [inline rpc/remote_signer_test.go]: Redundant extra line.

File: rpc/remote_signer.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+// RemoteSigner capable of signing requests by using
+// BLS secret keys retrieved from a keyvault.
+type RemoteSigner struct {
+	keyVault keyvault.Store
+}
+
+// NewRemoteSigner instantiates a new server instance using
+// a keyvault for retrieving.
+func NewRemoteSigner(ctx context.Context, keyVault keyvault.Store) *RemoteSigner {
```
Comment (farazdagi) [inline rpc/remote_signer.go:24]: Do we need to pass context here, it seems it is unused?

File: keyvault/deterministic/store.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+	).Info("Initialized deterministic keyvault")
+	s := &Store{
+		pubKeys:             pubKeys,
+		pubKeysToSecretKeys: pubKeysToSecretKeys,
+	}
+	return s, nil
+}
+
+// GetSecretKey returns the corresponding secret key for a BLS12-381 public key.
+func (s *Store) GetSecretKey(ctx context.Context, pubKey bls.PublicKey) (bls.SecretKey, error) {
```
Comment (farazdagi) [inline keyvault/deterministic/store.go:55]: Again, `ctx` is unused, not sure if that is intentional (to be used in further cleanup PR). The same for `GetPublicKeys()`

**Review by farazdagi — APPROVED** [review]

File: keyvault/deterministic/store.go (hunk 4 above)
Comment (rauljordan) [reply keyvault/deterministic/store.go:55]: It's important to always have ctx as the first argument in interface definitions as a best practice, such as:
```go
type Store interface {
  GetSecretKey(context.Context, bls.PublicKey) (bls.SecretKey, error)
}
```
As most implementations of the interface will wish to leverage context in one way or another

Comment (rauljordan) [reply keyvault/deterministic/store.go:55]: For example, a `Store` implementation using something such as aws S3 or hashicorp vault might need the incoming context for their own purposes

Comment (farazdagi) [reply keyvault/deterministic/store.go:55]: Yep, I also think that `ctx` is probably the only good case where you can add ~variable~parameter for future, not immediate use within a created function.

**Pull request merged**